### PR TITLE
feat(run/receive-auth): add run sample for receive-auth func

### LINKED ISSUE #2986: Sample Request: Receive authenticated requests
Request for a Go sample on [this documentation page](https://cloud.google.com/run/docs/authenticating/service-to-service#receive-request).

based on captured user feedback b/278311493

Only Python sample available: https://github.com/GoogleCloudPlatform/python-docs-samples/blob/HEAD/run/service-auth/receive.py

### PR DESCRIPTION
**Pull request by ryuji-cre8ive**
## Description
* check jwt with email from the request.
* almost same func as python one
  * https://github.com/GoogleCloudPlatform/python-docs-samples/blob/HEAD/run/service-auth/receive.py 

Fixes #2986
Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed [Contributing Guidelines from CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md)
- [x] **Tests** pass:   `go test -v ./..` (see [Testing](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#testing))
- [x] **Code formatted**:   `gofmt` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [x] **Vetting** pass:   `go vet` (see [Formatting](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/CONTRIBUTING.md#formatting))
- [x] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [x] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [x] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/golang-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

### CODE REVIEW

**Review by Mani8841 — APPROVED** [review]

**Comment by telpirion** [comment]
Make sure that you run `go vet` on the files in your PR.

File: run/receive-auth/main_test.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+		{
+			name:           "Invalid Bearer token",
+			header:         "Bearer " + "invalid-token",
+			wantStatusCode: http.StatusOK,
+			wantResponse:   "Unable to parse token: token contains an invalid number of segments",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
Comment (telpirion) [inline run/receive-auth/main_test.go:61]: issue: remove the call to `t.Run()` -- you don't need to run these tests in parallel.

File: run/receive-auth/main_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		})
+	}
+}
+
+func createToken(email string) string {
+	token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
+		"email": email,
+	})
+	fmt.Printf("token: %+v\n", token)
+	// ReWrite <my-secret-key> for your secret key in the same way to run/service-auth/receive.go
```
Comment (telpirion) [inline run/receive-auth/main_test.go:92]: typo: "rewrite"

**Review by telpirion — CHANGES_REQUESTED** [review]

**Review by telpirion — APPROVED** [review]

File: run/receive-auth/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    ca-certificates && \
+    rm -rf /var/lib/apt/lists/*
+
+# Copy the binary to the production image from the builder stage.
+COPY --from=builder /app/server /app/server
+
+# Run the web service on container startup.
+CMD ["/app/server"]
+
+# [END cloudrun_serviceauth_dockerfile]
```
Comment (grayside) [inline run/receive-auth/Dockerfile:51]: While most existing samples use Dockerfiles, the default should now be to rely on Cloud Run source deploys where there is no runtime customization necessary.

File: run/receive-auth/main.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+	if authType == "Bearer" {
+		ctx := r.Context()
+		config := &oauth2.Config{
+			ClientID:     "<your-client-id>",
+			ClientSecret: "<your-client-secret>",
+			Endpoint:     oauth2.Endpoint{
+				TokenURL: "https://oauth2.googleapis.com/token",
+			},
+		}
```
Comment (grayside) [inline run/receive-auth/main.go:71]: The Python example is not showing re-authenticating the token. Perhaps that should be done, but it exceeds current scope.

File: run/receive-auth/main.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+		token, err := config.TokenSource(ctx, &oauth2.Token{AccessToken: creds}).Token()
+		if err != nil {
+			http.Error(w, "Unable to parse token: "+err.Error(), http.StatusInternalServerError)
+			return
+		}
```
Comment (grayside) [inline run/receive-auth/main.go:77]: Using a TokenSource here seems confusing, all we want is to validate the JWT.

Maybe we should simplify to something like: 

```go
token := &oauth2.Token{AccessToken: creds})
if !token.Validate() {
```

**Review by grayside — CHANGES_REQUESTED** [review]
Thanks very much for this sample contribution! There are some opportunities to simplify this for future maintenance and stronger alignment with the python sample we're trying to match.

File: run/receive-auth/Dockerfile (hunk 3 above)
Comment (ryuji-cre8ive) [reply run/receive-auth/Dockerfile:51]: Thank you for contacting us.
Does this mean that you do not need a Dockerfile?

File: run/receive-auth/main.go (hunk 5 above)
Comment (ryuji-cre8ive) [reply run/receive-auth/main.go:77]: Thanks for letting me know.
Noted. I will amend it as such.

**Comment by telpirion** [comment]
Hi @ryuji-cre8ive -- it looks like this pull request still fails Lint, my last round of comments haven't been addressed, and the last interaction on this PR was over two months ago. Thus I will close this PR.

**Pull request closed without merging**